lib/Backend/arm64/LowerMD.cpp

```diff
@@ -6008,73 +5958,70 @@ LowererMD::GenerateClz(IR::Instr * instr)
 void
 LowererMD::SaveDoubleToVar(IR::RegOpnd * dstOpnd, IR::RegOpnd *opndFloat, IR::Instr *instrOrig, IR::Instr *instrInsert, bool isHelper)
 {
+    Assert(opndFloat->GetType() == TyFloat64);
+
     // Call JSNumber::ToVar to save the float operand to the result of the original (var) instruction
-    IR::Opnd * symVTableDst;
-    IR::Opnd * symDblDst;
-    IR::Opnd * symTypeDst;
-    IR::Instr *newInstr;
-    IR::Instr * numberInitInsertInstr = nullptr;
-    if (instrOrig->dstIsTempNumber)
-    {
-        // Use the original dst to get the temp number sym
-        StackSym * tempNumberSym = this->m_lowerer->GetTempNumberSym(instrOrig->GetDst(), instrOrig->dstIsTempNumberTransferred);
-
-        // LEA dst, &tempSym
-        IR::SymOpnd * symTempSrc = IR::SymOpnd::New(tempNumberSym, TyMachPtr, this->m_func);
-        newInstr = IR::Instr::New(Js::OpCode::LEA, dstOpnd, symTempSrc, this->m_func);
+
+    // s1 = MOV opndFloat
+    IR::RegOpnd *s1 = IR::RegOpnd::New(TyMachReg, m_func);
+    IR::Instr *mov = IR::Instr::New(Js::OpCode::FMOV, s1, opndFloat, m_func);
+    instrInsert->InsertBefore(mov);
+
+    if (m_func->GetJITFunctionBody()->IsAsmJsMode())
+    {
+        // s1 = FMOV src
+        // tmp = NOT s1
+        // tmp = AND tmp, 0x7FF0000000000000ull
+        // cbz tmp, helper
+        // jmp done
+        // helper:
+        // tmp2 = AND s1, 0x000FFFFFFFFFFFFFull
+        // cbz tmp2, done
+        // s1 = JavascriptNumber::k_Nan
+        // done:
+
```

Comment:
Here's an alternative:
s1 = FMOV_GEN src    ; assuming src is in an FP register
tmp = UBFX s1, #52, #11   ; extract exponent, bits 52-62
CMP tmp, #0x7ff    ; check for 0x7ff
BNE done    ; if not, done; otherwise fall through
tmp2 = UBFX s1, #0, #52   ; extract mantissa, bits 0-51
CBZ tmp2, done    ; if zero, done
s1 = JavascriptNumber::k_Nan    ; load NaN
done:

Reply:
During the Layout phase later on helper blocks are moved to the end of the function, so by treating the special case as a helper block the emitted code will take the shape(with the other corrections applied)

s1 = FMOV_GEN src
tmp = UBFX s1, #52, #11
cmp = tmp, 0x7FF
beq tmp, helper
done:
...
helper:
tmp2 = tmp2 = UBFX s1, #0, #52
cbz tmp2, done
s1 = JavascriptNumber::k_Nan

Reply:
Ah, didn't know that.